nifi-toolkit/nifi-toolkit-tls/src/main/java/org/apache/nifi/toolkit/tls/diagnosis/TlsToolkitGetDiagnosisStandalone.java

```diff
@@ -0,0 +1,664 @@
+            standalone.keystore = checkPasswordForKeystoreAndLoadKeystore(keystorePassword, standalone.keystorePath, standalone.keystoreType);
+            standalone.truststore = checkPasswordForKeystoreAndLoadKeystore(truststorePassword, standalone.truststorePath, standalone.truststoreType);
+            if (!(standalone.keystore == null) && !(standalone.truststore == null)) {
+                // TODO: Refactor "dangerous" logic to method which throws exceptions
+                KeyStore.PrivateKeyEntry privateKeyEntry = standalone.extractPrimaryPrivateKeyEntry(standalone.keystore, keystorePassword);
+                if (privateKeyEntry != null) {
+                    if (standalone.identifyHostUsingKeystore(privateKeyEntry)) {
+                        outputSummary.put(TRUSTSTORE, standalone.checkTruststore(privateKeyEntry));
+
+                        displaySummaryReport();
+                    } else {
+                        System.exit(-1);
+                    }
+                } else {
+                    System.exit(-1);
+                }
+            } else {
+                System.exit(-1);
+            }
+        } else {
+            System.exit(-1);
+        }
+    }
+
+    private KeyStore.PrivateKeyEntry extractPrimaryPrivateKeyEntry(KeyStore keystore, char[] keystorePassword) {
+        try {
+            KeyStore.PasswordProtection keystorePasswordProtection = new KeyStore.PasswordProtection(keystorePassword);
+            List<String> keystoreAliases = Collections.list(keystore.aliases());
+            Map<String, KeyStore.Entry> privateEntries = keystoreAliases.stream()
+                    .map(alias -> retrieveEntryFromKeystore(keystorePasswordProtection, alias))
+                    .filter(Objects::nonNull)
+                    .filter(t -> t.getValue() instanceof KeyStore.PrivateKeyEntry)
+                    .collect(Collectors.toMap(Tuple::getKey, Tuple::getValue));
+
+            //Check # of privateKeyEntry(s)
+            if (privateEntries.size() == 0) {
+                logger.error("No privateKeyEntry in keystore. Cannot explore keystore identification.");
+                return null;
+            } else if (privateEntries.size() > 1) {
+                logger.info("Keystore has multiple privateKeyEntries. Using the first privateKeyEntry in the list: " + new ArrayList<>(privateEntries.keySet()).get(0));
+                logger.warn("Recommended to have a single PrivateKeyEntry in keystore");
+                logger.warn("Available PrivateKeyEntries: " + StringUtils.join(privateEntries.keySet(), ", "));
+            } else {
+                logger.info("Keystore has single privateKeyEntry: " + new ArrayList<>(privateEntries.keySet()).get(0));
+            }
+            return ((KeyStore.PrivateKeyEntry) new ArrayList<>(privateEntries.values()).get(0));
+        } catch (KeyStoreException e) {
+            logger.error("Something went wrong: " + e.getLocalizedMessage());
```

Comment:
I'm not sure if I understand this one. I have added the KeystoreException **e** to the `logger.error` as `e.getLocalizedMessage()`

Reply:
Sorry for the lack of clarity, I recommend passing the exception object to the logger so that the stack trace is also logged:

```suggestion
            logger.error("Something went wrong: {}", e.getLocalizedMessage(), e);
```

Reply:
Yes I'll add that.